modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/internal/UncancelablePromise.kt

```diff
@@ -0,0 +1,95 @@
+package arrow.effects.internal
+
+import arrow.Kind
+import arrow.core.*
+import arrow.effects.Promise
+import arrow.effects.typeclasses.Async
+import java.util.concurrent.atomic.AtomicReference
+
+internal class UncancelablePromise<F, A>(AS: Async<F>) : Promise<F, A>, Async<F> by AS {
+
+  internal sealed class State<out A> {
+    data class Pending<A>(val joiners: List<(Either<Throwable, A>) -> Unit>) : State<A>()
+    data class Full<A>(val value: A) : State<A>()
```

Comment:
I was trying to go for a different AST naming altogether. Something that lies more in line with `Promise`, something like `Fulfilled` and `Unfulfilled` but I reverted back during the PR somewhere I think :D

Naming here is for our own benefit mostly since the state is internal to the internal implementations. I'll rename both to `Complete` to be more in line with the APIs.